### PR TITLE
`Paywalls`: new `PaywallFooterView` to replace `modes`

### PR DESCRIPTION
**Pull request by joshdholtz**
## Motivation

It was a little unclear how and when to use the different `PaywallView` modes (`.fullscreen`, `.card`, and `.condensedCard`). The actual functionality of the modes is great but they seemed like three different components in one view and it was just trying to do too much for one public API.

This change will...
- Remove modes from being a public parameter in `PaywallView` and only allow it to be full screen
- Create a new `PaywallFooterView` which will combine `.footer` (previously `.card`) and `.condensedFooter` (previously `.condensedCard`)
- Removed the `modes` parameter from `PaywallView`

### New `PaywallFooterView`

- Wraps `PaywallView` and will only choose between `.footer` and `.footerCondensed`
- Currently offers a `condensed` parameter but this might be switched to being a remote setting inside of `PaywalData`

### New `.paywallFooter()` view modifier

- Easy way for developers to place `PaywallFooterView` where it is supposed to be placed
- This places the view in `.safeAreaInset(edge: .bottom)`
  - This works on any view but is ideal in `ScrollView` as it will automatically handle adjusting the scroll insets with the paywalls size

### New footer structure and animation (with some hacks)

- No parts of `PaywallFooterView` uses `.overlay` anymore
  - The whole grows in `.condensedFooter` when "All Plans" is toggled

### My sample app

https://github.com/RevenueCat/purchases-ios/assets/401294/17c52bba-956d-404e-b991-a2272e7fc4f1

### Simple sample app 

https://github.com/RevenueCat/purchases-ios/assets/401294/4ea94156-8367-4dea-9640-0cce8ffa1deb

### CODE REVIEW

File: RevenueCatUI/PaywallFooterView.swift [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            condensed: condensed,
+            fonts: fonts,
+            introEligibility: .default(),
+            purchaseHandler: .default()
+        )
+    }
+
+    init(
+        offering: Offering?,
+        condensed: Bool = false,
```
Comment (joshdholtz) [inline RevenueCatUI/PaywallFooterView.swift:69]: Going with `condensed: Bool` for now to make it easier on which type of footer is wanted but this might get moved to a remote configuration in `PaywallData`

File: RevenueCatUI/Data/Constants.swift [hunk 2]
```diff
@@ -12,10 +12,12 @@ enum Constants {
 
     static let defaultAnimation: Animation = .easeInOut(duration: 0.2)
     static let fastAnimation: Animation = .easeInOut(duration: 0.1)
-    static let displayAllPlansAnimation: Animation = .easeInOut(duration: 0.2)
+    static let toggleAllPlansAnimation: Animation = .spring(response: 0.35, dampingFraction: 0.7)
 
     static let defaultCornerRadius: CGFloat = 20
 
+    static let maxFooterHeight: CGFloat = 400
```
Comment (joshdholtz) [inline RevenueCatUI/Data/Constants.swift:19]: Leaving unlimited height for footer felt bad so just putting this at 400 for now? But not sold on this at all 🙃

File: RevenueCatUI/Data/Constants.swift [hunk 3]
```diff
@@ -12,10 +12,12 @@ enum Constants {
 
     static let defaultAnimation: Animation = .easeInOut(duration: 0.2)
     static let fastAnimation: Animation = .easeInOut(duration: 0.1)
-    static let displayAllPlansAnimation: Animation = .easeInOut(duration: 0.2)
+    static let toggleAllPlansAnimation: Animation = .spring(response: 0.35, dampingFraction: 0.7)
```
Comment (NachoSoto) [inline RevenueCatUI/Data/Constants.swift:15]: 😍

File: RevenueCatUI/Modifiers/FooterHidingModifier.swift [hunk 4]
```diff
@@ -41,27 +38,12 @@ private struct FooterHidingModifier: ViewModifier {
...
-                        )
-                        .frame(height: self.hide ? 0 : nil)
-                        .blur(radius: self.hide ? Self.blurRadius : 0)
-                }
+            content
+                .onSizeChange(.vertical) { self.height = $0 }
+                .opacity(self.hide ? 0 : 1)
+                .frame(height: self.hide ? 0 : nil, alignment: .top)
+                .clipped()
+                .transition(.move(edge: .bottom))
```
Comment (NachoSoto) [inline RevenueCatUI/Modifiers/FooterHidingModifier.swift:46]: This is so much simpler

File: RevenueCatUI/Templates/Template2View.swift [hunk 5]
```diff
@@ -83,20 +86,35 @@ struct Template2View: TemplateViewType {
...
+            VStack {
+                self.packagesScrollHack
+
                 Spacer()
-            } else {
-                self.packages
-                    .onSizeChange(.vertical) { if $0 > 0 { self.containerHeight = $0 } }
-                    .hideFooterContent(self.configuration,
-                                       hide: !self.displayingAllPlans,
-                                       offset: self.containerHeight)
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/Template2View.swift:94]: We can remove this property now, right?

File: RevenueCatUI/Templates/Template2View.swift [hunk 6]
```diff
@@ -83,20 +86,35 @@ struct Template2View: TemplateViewType {
...
             }
+            // Sets a max footer height so there is enough room for paywall content above
+            .frame(maxHeight: configuration.mode == .fullScreen ? nil : Constants.maxFooterHeight)
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/Template2View.swift:95]: Why not use `.infinite`? I think it accomplishes the same goal without hardcoding a magic number.

File: RevenueCatUI/Templates/Template2View.swift [hunk 7]
```diff
@@ -83,20 +86,35 @@ struct Template2View: TemplateViewType {
...
         .frame(maxHeight: .infinity)
     }
 
+    /// Hack because scrollableIfNecessary() doesn't work on iOS 15 because of unknown reason
+    private var packagesScrollHack: some View {
+        Group {
+            if #available(iOS 16.0, macOS 13.0, tvOS 16.0, watchOS 9.0, *) {
+                self.packages
+                    // Another hack because bounce animation temporarily clips top package a little bits
+                    .padding(.top, 5)
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/Template2View.swift:108]: :( let's look at a better way of doing this.

File: RevenueCatUI/Templates/Template2View.swift [hunk 8]
```diff
@@ -83,20 +86,35 @@ struct Template2View: TemplateViewType {
...
+            .frame(maxHeight: configuration.mode == .fullScreen ? nil : Constants.maxFooterHeight)
+            .hideFooterContent(self.configuration,
+                               hide: !self.configuration.mode.shouldDisplayPackages && !self.displayingAllPlans)
         }
         .frame(maxHeight: .infinity)
     }
 
+    /// Hack because scrollableIfNecessary() doesn't work on iOS 15 because of unknown reason
+    private var packagesScrollHack: some View {
+        Group {
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/Template2View.swift:104]: The recommended way of doing this is:
```suggestion
    @ViewBuilder
    private var packagesScrollHack: some View {
```

File: RevenueCatUI/View+PresentPaywallFooter.swift [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    /// - Note: If loading the `CustomerInfo` fails (for example, if Internet is offline),
+    /// the paywall won't be displayed.
+    ///
+    /// ### Related Articles
+    /// [Documentation](https://rev.cat/paywalls)
+    public func paywallFooter(
+        condensed: Bool = false,
+        fonts: PaywallFontProvider = DefaultPaywallFontProvider(),
+        purchaseCompleted: PurchaseCompletedHandler? = nil
+    ) -> some View {
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywallFooter.swift:32]: 👏🏻  I think this is great.

File: RevenueCatUI/View+PresentPaywallFooter.swift [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+    /// Presents a ``PaywallFooterView`` at the bottom of a view with the given offering.
+    /// ```swift
+    /// var body: some View {
+    ///    YourPaywall()
+    ///      .paywallFooter(offering: offering)
+    /// }
+    /// ```
+    /// - Note: If loading the `CustomerInfo` fails (for example, if Internet is offline),
+    /// the paywall won't be displayed.
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywallFooter.swift:50]: Ditto

File: RevenueCatUI/View+PresentPaywallFooter.swift [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                introEligibility: introEligibility,
+                purchaseHandler: purchaseHandler
+            ))
+    }
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
+@available(macOS, unavailable)
+@available(tvOS, unavailable)
+struct PresentingPaywallFooterModifier: ViewModifier {
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywallFooter.swift:92]: This can be `private`.

File: Tests/RevenueCatUITests/PaywallFooterTests.swift [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+@MainActor
+class PaywallFooterTests: TestCase {
+
+    override func setUpWithError() throws {
+        try super.setUpWithError()
+
+        try AvailabilityChecks.iOS16APIAvailableOrSkipTest()
+    }
+
+    func testPresentWithPurchaseHandler() throws {
```
Comment (NachoSoto) [inline Tests/RevenueCatUITests/PaywallFooterTests.swift:26]: Cool.
Let's add a snapshot test too.

**Review by NachoSoto — COMMENTED** [review]
I haven't tested this, but assuming it hasn't broken any of the permutations (iOS / template / etc) this is awesome 👏🏻 
- Much better API
- Simpler implementation
- More discoverable
- More intuitive to use

One thing is that we can make `PaywallViewMode` `internal`, and "hide" the parameter from the `PaywallView` public constructors, right?

File: RevenueCatUI/PaywallFooterView.swift [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+/// A SwiftUI view for displaying a `PaywallData` for an `Offering`.
+///
+/// ### Related Articles
+/// [Documentation](https://rev.cat/paywalls)
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
+@available(watchOS, unavailable, message: "RevenueCatUI does not support watchOS yet")
+@available(macOS, unavailable, message: "RevenueCatUI does not support macOS yet")
+@available(tvOS, unavailable, message: "RevenueCatUI does not support tvOS yet")
+@available(macCatalyst, unavailable, message: "RevenueCatUI does not support Catalyst yet")
+public struct PaywallFooterView: View {
```
Comment (NachoSoto) [inline RevenueCatUI/PaywallFooterView.swift:20]: Now that I think about it, does this need to be `public`?

File: RevenueCatUI/PaywallFooterView.swift [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+    private let mode: PaywallViewMode
+    private let fonts: PaywallFontProvider
+    private let introEligibility: TrialOrIntroEligibilityChecker?
+    private let purchaseHandler: PurchaseHandler?
+
+    @State
+    private var offering: Offering?
+    @State
+    private var error: NSError?
```
Comment (NachoSoto) [inline RevenueCatUI/PaywallFooterView.swift:30]: This isn't used.

File: RevenueCatUI/PaywallFooterView.swift [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+@available(tvOS, unavailable, message: "RevenueCatUI does not support tvOS yet")
+@available(macCatalyst, unavailable, message: "RevenueCatUI does not support Catalyst yet")
+public struct PaywallFooterView: View {
+
+    private let mode: PaywallViewMode
+    private let fonts: PaywallFontProvider
+    private let introEligibility: TrialOrIntroEligibilityChecker?
+    private let purchaseHandler: PurchaseHandler?
+
+    @State
```
Comment (NachoSoto) [inline RevenueCatUI/PaywallFooterView.swift:27]: This shouldn't be `@State`.

File: RevenueCatUI/Templates/Template2View.swift [hunk 16]
```diff
@@ -41,7 +41,10 @@ struct Template2View: TemplateViewType {
             Spacer()
 
             self.scrollableContent
-                .scrollableIfNecessary()
+                // Disabling in overlay mode here because of animation issues with ViewThatFits
+                // Scrolling is enabled in overlay mode in the packages
+                // Bonus is that CTA stays visible and doesn't scroll
+                .scrollableIfNecessary(enabled: configuration.mode == .fullScreen)
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply RevenueCatUI/Templates/Template2View.swift:47]: Yeah I don't think we can do this, this probably breaks with larger dynamic types.

File: RevenueCatUI/Templates/Template2View.swift [hunk 17]
```diff
@@ -83,20 +86,35 @@ struct Template2View: TemplateViewType {
...
-                self.packages
-                    .onSizeChange(.vertical) { if $0 > 0 { self.containerHeight = $0 } }
-                    .hideFooterContent(self.configuration,
-                                       hide: !self.displayingAllPlans,
-                                       offset: self.containerHeight)
             }
+            // Sets a max footer height so there is enough room for paywall content above
+            .frame(maxHeight: configuration.mode == .fullScreen ? nil : Constants.maxFooterHeight)
+            .hideFooterContent(self.configuration,
+                               hide: !self.configuration.mode.shouldDisplayPackages && !self.displayingAllPlans)
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/Template2View.swift:97]: Why this change?

File: RevenueCatUI/PaywallFooterView.swift (hunk 13 above)
Comment (NachoSoto) [reply RevenueCatUI/PaywallFooterView.swift:20]: Changed.

File: RevenueCatUI/PaywallFooterView.swift (hunk 15 above)
Comment (NachoSoto) [reply RevenueCatUI/PaywallFooterView.swift:27]: Changed.

File: RevenueCatUI/PaywallFooterView.swift (hunk 14 above)
Comment (NachoSoto) [reply RevenueCatUI/PaywallFooterView.swift:30]: Removed

File: RevenueCatUI/PaywallFooterView.swift (hunk 1 above)
Comment (NachoSoto) [reply RevenueCatUI/PaywallFooterView.swift:69]: Nah this makes sense, I think this should be configured from the SDK as it depends on how it's used in the app.

File: RevenueCatUI/Templates/Template2View.swift (hunk 5 above)
Comment (NachoSoto) [reply RevenueCatUI/Templates/Template2View.swift:94]: Removed.

File: RevenueCatUI/Templates/Template2View.swift (hunk 6 above)
Comment (NachoSoto) [reply RevenueCatUI/Templates/Template2View.swift:95]: Removed this. We can add a parameter for the maximum height if users request it.

File: RevenueCatUI/PaywallFooterView.swift [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+/// A SwiftUI view for displaying a `PaywallData` for an `Offering`.
+///
+/// ### Related Articles
+/// [Documentation](https://rev.cat/paywalls)
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
+@available(watchOS, unavailable, message: "RevenueCatUI does not support watchOS yet")
+@available(macOS, unavailable, message: "RevenueCatUI does not support macOS yet")
+@available(tvOS, unavailable, message: "RevenueCatUI does not support tvOS yet")
+@available(macCatalyst, unavailable, message: "RevenueCatUI does not support Catalyst yet")
+internal struct PaywallFooterView: View {
```
Comment (NachoSoto) [inline RevenueCatUI/PaywallFooterView.swift:20]: Made this `internal` now.

**Review by NachoSoto — APPROVED** [review]
